fili-core/src/main/java/com/yahoo/bard/webservice/web/handlers/CacheWeightCheckRequestHandler.java

```diff
@@ -0,0 +1,225 @@
+            final DataApiRequest request,
+            final DruidAggregationQuery<?> druidQuery,
+            final ResponseProcessor response
+    ) {
+        // Heuristic test to let requests with very low estimated cardinality directly through
+        if (queryWeightUtil.skipWeightCheckQuery(druidQuery)) {
+            return next.handleRequest(context, request, druidQuery, response);
+        }
+
+        BardQueryInfo.getBardQueryInfo().incrementCountWeightCheck();
+        final WeightCheckResponseProcessor weightCheckResponse =
+                new WeightCheckResponseProcessor(response);
+        final DruidAggregationQuery<?> weightEvaluationQuery = queryWeightUtil.makeWeightEvaluationQuery(druidQuery);
+        Granularity granularity = druidQuery.getInnermostQuery().getGranularity();
+        final long queryRowLimit = queryWeightUtil.getQueryWeightThreshold(granularity);
+
+        try {
+            LOG.debug("Weight query {}", writer.writeValueAsString(weightEvaluationQuery));
+        } catch (JsonProcessingException e) {
+            LOG.warn("Weight Query json exception:", e);
+        }
+
+        CacheService cacheService = new CacheService();
+        String cacheKey = null;
+        try {
+            cacheKey = getKey(druidQuery, mapper, writer);
+        } catch (JsonProcessingException e) {
+            LOG.warn("Cache key cannot be built: ", e);
+        }
+
+        WeightCheckRequestHandler weightCheckRequestHandler =
+                new WeightCheckRequestHandler(next, webService, queryWeightUtil, mapper);
+        SuccessCallback weightQuerySuccess = null;
+        // Construct success callback based on cache hit or cache miss
+        if (context.isReadCache()) {
+            String cacheResponse =
+                    cacheService.readCache(context, dataCache, querySigningService, druidQuery, cacheKey);
+            // Cache hit
+            if (cacheResponse != null) {
+                weightQuerySuccess = buildCacheSuccessCallback(
+                        context,
+                        request,
+                        druidQuery,
+                        weightCheckResponse,
+                        queryRowLimit,
+                        cacheKey,
+                        cacheService,
+                        weightCheckRequestHandler,
+                        false
+                );
+            }
+        } else {
+            // Write to cache
+            weightQuerySuccess = buildCacheSuccessCallback(
+                    context,
+                    request,
+                    druidQuery,
+                    weightCheckResponse,
+                    queryRowLimit,
+                    cacheKey,
+                    cacheService,
+                    weightCheckRequestHandler,
+                    true
+            );
+        }
+        HttpErrorCallback error = response.getErrorCallback(druidQuery);
+        FailureCallback failure = response.getFailureCallback(druidQuery);
```

Comment:
these guys can move up because we'll need them in every exit point.
Also, in most cases, we'll need to wrap the super success callback, so we can add:
```suggestion
        HttpErrorCallback error = response.getErrorCallback(druidQuery);
        FailureCallback failure = response.getFailureCallback(druidQuery);
        SuccessCallback classicCallback = super.buildSuccessCallback(
                context,
                request,
                druidQuery,
                weightCheckResponse,
                queryRowLimit
        );

        SuccessCallback cahingSuccessCallback = buildCacheSuccessCallback(
                classicCallback,
                cacheService,
                druidQuery,
                weightCheckResponse
        );
```